wfe2/wfe.go

```diff
@@ -2119,6 +2119,83 @@ func (wfe *WebFrontEndImpl) refundNewOrderLimits(ctx context.Context, transactio
 	}
 }
 
+// orderMatchesReplacement checks if the order matches the provided certificate
+// as identified by the provided ARI CertID. This function ensures that:
+//   - the certificate being replaced exists,
+//   - the requesting account owns that certificate, and
+//   - the names in this new order match the names in the certificate
+//     being replaced.
+func (wfe *WebFrontEndImpl) orderMatchesReplacement(ctx context.Context, acct *core.Registration, names []string, serial string) (*probs.ProblemDetails, error) {
+	oldCert, err := wfe.sa.GetCertificate(ctx, &sapb.Serial{Serial: serial})
+	if err != nil {
+		if errors.Is(err, berrors.NotFound) {
+			return probs.NotFound("Existing certificate could not found"), err
+		}
+		return probs.ServerInternal("Failed to retrieve existing certificate"), err
+	}
+
+	if oldCert.RegistrationID != acct.ID {
+		return probs.Unauthorized("Requester account did request the certificate being replaced by this order"), nil
+	}
+	parsedCert, err := x509.ParseCertificate(oldCert.Der)
+	if err != nil {
+		return probs.ServerInternal("Error parsing certificate replaced by this order"), err
+	}
+
+	// While draft-ietf-acme-ari-02 calls for the new order to "share a
+	// preponderance of identifiers" with the certificate being replaced, our
+	// implementation will be more lenient, requiring just one.
+	var namesMatch bool
+	for _, name := range names {
+		if parsedCert.VerifyHostname(name) == nil {
+			namesMatch = true
+			break
+		}
+	}
+	if !namesMatch {
+		return probs.Malformed("Certificate replaced by this order does not have matching identifiers"), nil
+	}
+	return nil, nil
+}
+
+func (wfe *WebFrontEndImpl) determineARIWindow(ctx context.Context, serial string) (core.RenewalInfo, *probs.ProblemDetails, error) {
+	// Check if the serial is impacted by an incident.
+	result, err := wfe.sa.IncidentsForSerial(ctx, &sapb.Serial{Serial: serial})
+	if err != nil {
+		return core.RenewalInfo{}, web.ProblemDetailsForError(err, "Checking if existing certificate is impacted by an incident"), err
+	}
+
+	if len(result.Incidents) > 0 {
+		// The existing certificate is impacted by an incident, renew immediately.
+		return core.RenewalInfoImmediate(wfe.clk.Now()), nil, nil
+	}
+
+	// Check if the serial is revoked.
+	status, err := wfe.sa.GetCertificateStatus(ctx, &sapb.Serial{Serial: serial})
+	if err != nil {
+		return core.RenewalInfo{}, web.ProblemDetailsForError(err, "Checking if existing certificate has been revoked"), err
+	}
+
+	if status.Status == string(core.OCSPStatusRevoked) {
+		// The existing certificate is revoked, renew immediately.
+		return core.RenewalInfoImmediate(wfe.clk.Now()), nil, nil
+	}
+
+	// It's okay to use GetCertificate (vs trying to get a precertificate),
+	// because we don't intend to serve ARI for certs that never made it past
+	// the precert stage.
+	cert, err := wfe.sa.GetCertificate(ctx, &sapb.Serial{Serial: serial})
```

Comment:
Consider using the same text as https://github.com/letsencrypt/boulder/pull/7298/files#diff-3f950e720c205ce9fa8dea12c6fd7fd44272c2671f19d0e06962abfbea00d491R2129-R2135